### PR TITLE
feat(pulsar): message keys as config field

### LINKED ISSUE #13117: Support Pulsar keys in topics
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

An application with several deployment push its logs on a Pulsar topic.
So, we have 1 app, N deployments, 1 topic.

It's then easier to filter on messages keys or keep ordering of messages

See https://pulsar.apache.org/docs/2.3.0/concepts-messaging/#ordering-guarantee

### Attempted Solutions

-

### Proposal

Add specific key like `_pulsar_key` with remap transformer

Or 

Add sink config property `key` filled with an evaluation of a VRL

### References

_No response_

### Version

0.21.2

### PR DESCRIPTION
**Pull request by miton18**
Fix #13117

### CODE REVIEW

File: src/sinks/pulsar.rs [hunk 1]
```diff
@@ -59,6 +58,9 @@ pub struct PulsarSinkConfig {
         skip_serializing_if = "crate::serde::skip_serializing_if_default"
     )]
     pub acknowledgements: AcknowledgementsConfig,
+
+    /// Log field to use as Pulsar message key
+    key_field: Option<String>,
```
Comment (fuchsnj) [inline src/sinks/pulsar.rs:63]: maybe this should be named `partition_key_field` since there is also an "ordering key" which could be confused with this?

**Review by fuchsnj — COMMENTED** [review]
Overall this looks good.
Please update the docs with this new field (`website/cue/reference/components/sinks/pulsar.cue`)

File: src/sinks/pulsar.rs (hunk 1 above)
Comment (miton18) [reply src/sinks/pulsar.rs:63]: From my understanding, there is only 1 kind of key on a message. This key is used for ordering and consumption.

```rust
// pulsar-rs/producer.rs
    pub fn with_partition_key<S: Into<String>>(mut self, partition_key: S) -> Self {
        self.partition_key = Some(partition_key.into());
        self
    }
    /// sets the message's ordering key for key_shared subscription
    pub fn with_ordering_key<S: Into<String>>(mut self, partition_key: S) -> Self {
        self.partition_key = Some(partition_key.into());
        self
    }

    /// sets the message's partition key
    ///
    /// this is the same as `with_partition_key`, this method is added for
    /// more consistency with other clients
    pub fn with_key<S: Into<String>>(mut self, partition_key: S) -> Self {
        self.partition_key = Some(partition_key.into());
        self
    }
 ```
 But I can rename it `partition_key_field`

Comment (fuchsnj) [reply src/sinks/pulsar.rs]: That's interesting. I was just looking at the crate documentation, so I didn't realize it's setting the same value. In that case, I'm fine with the current naming.

**Comment by miton18** [comment]
I'm discovering cue scripts, thanks @fuchsnj

**Comment by fuchsnj** [comment]
Thanks @miton18 for your contribution!

**Pull request merged**